R/derive_last_dose.R

```diff
@@ -83,143 +94,147 @@
-      tmp_analysis_date = convert_dtc_to_dtm(
-        dtc = !!analysis_date,
+      )
+    )
+
+   dataset_tmp <- dataset %>%
+    mutate(
+      tmp_analysis_date = convert_date_to_dtm(
+        dt = !!analysis_date,
         date_imputation = NULL,
         time_imputation = "23:59:59"
       )
-    ) %>%
-    group_by(!!!by_vars, !!dataset_seq_var)
-
-  # if no traceability variables are required, simply calculate the last dose date
-  if (is.null(traceability_vars)) {
-    res <- res %>%
-      summarise(
-        ldose_idx = compute_ldose_idx(dose_end = .data$tmp_dose_end_date,
-                                      analysis_date = .data$tmp_analysis_date),
-        !!new_var := as.POSIXct(as.character(.data$tmp_dose_end_date[.data$ldose_idx]),
-                                tz = lubridate::tz(.data$tmp_dose_end_date))
-      ) %>%
-      ungroup()
-  } else {
-    # calculate the last dose date and get the appropriate traceability variables
-    res <- res %>%
-      mutate(
-        ldose_idx = compute_ldose_idx(dose_end = .data$tmp_dose_end_date,
-                                      analysis_date = .data$tmp_analysis_date),
-        !!new_var := as.POSIXct(as.character(.data$tmp_dose_end_date[.data$ldose_idx]),
-                                tz = lubridate::tz(.data$tmp_dose_end_date))
-      ) %>%
-      mutate_at(trace_vars_str, list(~ .[.data$ldose_idx])) %>%
-      distinct(!!new_var, !!!syms(trace_vars_str)) %>%
-      ungroup()
-  }
+    )
 
-  # return either date or date-time variable
-  if (!output_datetime) {
-    res <- mutate(res, !!new_var := as.Date(!!new_var))
-  }
+  # join datasets and keep unique last dose records (where dose_date is before or on analysis_date)
+  # for each by_var and dataset_seq_var
+  res <- dataset_tmp %>%
+    inner_join(dataset_ex, by = by_vars_str) %>%
+    group_by(!!!by_vars, !!dataset_seq_var) %>%
+    mutate(
+      tmp_ldose_idx = compute_ldose_idx(dose_date = .data$tmp_dose_date,
+                                        analysis_date = .data$tmp_analysis_date),
+      tmp_ldose_dt = as.POSIXct(as.character(.data$tmp_dose_date[.data$tmp_ldose_idx]),
+                                tz = lubridate::tz(.data$tmp_dose_date))
+    ) %>%
+    ungroup() %>%
+    filter(tmp_dose_date == tmp_ldose_dt) %>%
+    filter_extreme(by_vars = c(by_vars, vars(!!dataset_seq_var, tmp_dose_date)),
+                   order = c(by_vars, vars(!!dataset_seq_var, tmp_dose_date), dose_id),
+                   mode = "last") %>%
+    select(!!!by_vars, !!dataset_seq_var, !!!syms(ex_keep_vars_str), !!!syms(trace_vars_str))
 
-  # return dataset with additional column
-  left_join(dataset,
-            distinct(res,
-                     !!!by_vars, !!dataset_seq_var, !!new_var,
-                     !!!syms(trace_vars_str)),
+  # return original dataset and add EX source variables from last dose record
+  left_join(dataset, res,
             by = c(by_vars_str, as_string(quo_get_expr(dataset_seq_var))))
```

Comment:
The `inner_join()` and `filter()` statements only keep the observations from `dataset` that have an associated last dose from `dataset_ex`. Then the `left_join()` brings back the remaining observations and variables. However, with the changes you proposed using `filter_extreme()`, I agree the additional `left_join()` is no longer needed. Thanks for your suggestion!

Reply:
Actually I tested around the function after the below updates and realized if we remove the final `left_join()` this would only keep observations from input dataset that satisfy the filter condition `filter(is.na(tmp_dose_date) | is.na(tmp_analysis_date) | tmp_dose_date <= tmp_analysis_date)`. So I kept the `left_join()` to retain all observations from input dataset.

Reply:
Updated, thanks for your suggestions!